### ISSUE TITLE
A speedrun on consumer grade cards?

### ISSUE THREAD

**Issue by fzyzcjy** [KellerJordan/modded-nanogpt#29]
Hi thanks for the great repo! I would appreciate it if there can be a speed run on consumer cards e.g. RTX4090. Since it is 125M params, the RTX4090's 24GB memory should fit in the classical way, and thus it is trainable.

**Comment by KellerJordan** [comment]
a suggestion, to reduce memory you could run with a lower sequence length

**Comment by fzyzcjy** [comment]
I think so, thanks :) Just wondering whether there will be a speedrun like the current great one but focus on RTX4090's time, because many more people have consumer grade cards than H100s.

**Comment by naoro** [comment]
I think a Google colab speed runs would be also awesome -
That would greatly commodotize research and experimentation

**Comment by fzyzcjy** [comment]
That looks interesting! (But I guess it may be too hard to have it run in acceptable times...)

**Comment by alexjc** [comment]
Realistically single-card speed-run would need a smaller model too, otherwise it's too slow to experiment.

Thinking:
```
n_layer = 8
n_embd = 512
```

The sequence length during training has been a variable factor in the last speed run; for evaluation it's fine if it's the whole document clamped at maximum 1024 window size?

**Comment by fzyzcjy** [comment]
It seems H100 has 2000 TFLOPS for bf16 tensor core, while 4090 is about 330 TFLOPS. Thus 8xH100 5minute = 1x4090 4hour, which looks not bad!

The major problem looks like the memory is only 24GB... So we may not be able to do some optimizations.

**Comment by alexjc** [comment]
4h feels quite high for a speed run. Too hard to test ideas, no?

Working on some memory optimizations now, should help a lot...

**Comment by fzyzcjy** [comment]
Surely faster would be great! But if impossible, then 4h is better than nothing :(

**Comment by naoro** [comment]
> It seems H100 has 2000 TFLOPS for bf16 tensor core, while 4090 is about 330 TFLOPS. Thus 8xH100 5minute = 1x4090 4hour, which looks not bad!
> 
> The major problem looks like the memory is only 24GB... So we may not be able to do some optimizations.

A100 has got 40GB, cost is about 10usd for ~12 hours with about the same tflops as 4090.
So about 3.3usd per run. Not bad, I'd think.

**Comment by fzyzcjy** [comment]
If we optimize for cost, then 4090 is much cheaper than A100 per hour while having same tflops. Thus as long as we manage to fit in 24GB, then maybe we can further scale down the cost.

**Comment by KellerJordan** [comment]
A note: The current cost per run on an 8xH100 is about $1.90 (since it's about $3/hr for SXM H100s)

Personally, when I don't feel like spending that much, I go back to [speedrunning CIFAR-10](https://github.com/KellerJordan/cifar10-airbench). But I understand that might not be so interesting to everyone

**Comment by fzyzcjy** [comment]
Looks like 4090 is about $0.3/hr, thus 4hr = $1.2, which looks a bit cheaper. And moreover, some people have 4090s bought in their house (e.g. many people in r/LocalLlama, me, etc), while it seems less people buy A100/H100 in their house, and the bought cost is much cheaper than a cloud.

**Comment by lapp0** [comment]
I'm also interested in this variant.

Considering the long runtime, perhaps it makes sense to compete to minimize validation loss within a 1 hour run?

**Comment by KellerJordan** [comment]
I would guess that halving the sequence length (and going to batch size 16) will allow fitting the run into 24G memory, without impacting performance very much

**Comment by lapp0** [comment]
I achieved < 3.28 in a little under two hours with a few tweaks.

@KellerJordan are you interested in hosting a 1x4090 variant of the competition in this repo? If so, I'll submit a PR.

**Comment by fzyzcjy** [comment]
@lapp0 That looks great - $0.3/hr x 2hr = $0.6, which is 3x cheaper than $1.9 (8xH100)!

**Comment by KellerJordan** [comment]
I'd prefer to allow experimentation on 4090s, but still time the final speedrun on 8xH100, like it is now. I'm happy to help with timing for 4090 runs that look promising. That way, the benchmark doesn't encourage techniques which are specific to 1x4090 (e.g., using a much smaller batch size).

**Comment by lapp0** [comment]
Ah, without smaller batch size I get 2 hours 10 minutes.

https://gist.github.com/lapp0/2740a03a637ec926cf0eea90e541a0a6

The only changes necessary for 230 minute run which is effectively identical to the 8xH100 are
- `batch_size: int = 8`
- `sequence_length: int = 32 * 1024`
- `y = flex_attention(..., kernel_options={"BLOCK_M": 64, "BLOCK_N": 64, "BLOCK_M1": 32, "BLOCK_N1": 64, "BLOCK_M2": 64, "BLOCK_N2": 32}`
- update `run.sh` with `--nproc_per_node=1`

You can move this further by setting batch_size to 1, and tweaking the learning rates.

**Comment by fzyzcjy** [comment]
> however I'm really interested in a 4090 competition variant. Please let me know if someone is interested in hosting, otherwise I might create a fork myself :)

I am also quite interested in it and happy to host it and see it being improved :)

**Comment by banyan-god** [comment]
I was able to reproduce it on 4x 4090  31 minutes

**Comment by banyan-god** [comment]
> A note: The current cost per run on an 8xH100 is about $1.90 (since it's about $3/hr for SXM H100s)
> 
> Personally, when I don't feel like spending that much, I go back to [speedrunning CIFAR-10](https://github.com/KellerJordan/cifar10-airbench). But I understand that might not be so interesting to everyone

@KellerJordan  where do you rent your 8xH100

**Comment by LakshyAAAgrawal** [comment]
> Ah, without smaller batch size I get 2 hours 10 minutes.
> 
> https://gist.github.com/lapp0/2740a03a637ec926cf0eea90e541a0a6
> 
> The only changes necessary for 130 minute run which is effectively identical to the 8xH100 are
> 
>     * `batch_size: int = 16`
> 
>     * `sequence_length: int = 32 * 1024`
> 
>     * `y = flex_attention(..., kernel_options={"BLOCK_M": 64, "BLOCK_N": 64, "BLOCK_M1": 32, "BLOCK_N1": 64, "BLOCK_M2": 64, "BLOCK_N2": 32}`
> 
>     * update `run.sh` with `--nproc_per_node=1`
> 
> 
> You can improve runtime to 90 minutes while deviating from 8xH100 learning dynamics by setting batch_size to 1, and tweaking the learning rates.
> 
> I'll submit a PR for this setting as an experimentation tool, however I'm really interested in a 4090 competition variant. Please let me know if someone is interested in hosting, otherwise I might create a fork myself :)

Hey, can you please point me to whether 2 different runs, one with bsz=8, another with bsz=16 is still comparable in terms of numbers of tokens seen during training, everything else being fixed?

**Comment by lapp0** [comment]
>Hey, can you please point me to whether 2 different runs, one with bsz=8, another with bsz=16 is still comparable in terms of numbers of tokens seen during training, everything else being fixed?

This is true if you split the sequence length respectively. e.g. these runs are all equivalent:

H100:
```
torchrun --standalone --nproc_per_node=8 train_gpt2.py \
    --train.batch_size 8 --train.sequence_length 65536  # default values
```

4090:
```
torchrun --standalone --nproc_per_node=8 train_gpt2.py --gpt.flex_kernel_consumer True \
    --train.batch_size 16 --train.sequence_length --train.sequence_length 32768
```

The only caveat is that for each sequence in the batch, one sample is split. This occurs twice as frequently when the sequence length is halved.

**Comment by mosamdabhi** [comment]
> A note: The current cost per run on an 8xH100 is about $1.90 (since it's about $3/hr for SXM H100s)
> 
> Personally, when I don't feel like spending that much, I go back to [speedrunning CIFAR-10](https://github.com/KellerJordan/cifar10-airbench). But I understand that might not be so interesting to everyone

@KellerJordan Have you thought about doing something for ImageNet1K for quickly doing the same checks that you are doing but for vision?

**Comment by i90007** [comment]
> Hi thanks for the great repo! I would appreciate it if there can be a speed run on consumer cards e.g. RTX4090. Since it is 125M params, the RTX4090's 24GB memory should fit in the classical way, and thus it is trainable.

I'm currently trying to beat the validation loss result (3.28) on a free T4 graphics card, using Google Colab))

**Comment by MarktHart** [comment]
FYI: the code on master requires a tiny change to make it working, but runs in ~41 minutes on a 2xRTX4090 machine

**Comment by timmytonga** [comment]
@MarktHart Thanks for sharing your change. I am trying to run this on a 1x4090 gig. For the accumulation steps, don't you have to renormalize the loss before backward?

**Comment by MarktHart** [comment]
@timmytonga I think you might be right, or that there is at least an issue in that direction. The reason why I am not directly convinced is because `gradient_accumulation_steps=6`, which would increase the loss by 6x, and therefore effectively the LR by 6, right? It works too good right now for a factor 6 LR difference.

I am leaning towards the grad reduction playing a role with `dist.all_reduce(param.grad, op=dist.ReduceOp.AVG)`. So that would be `sum(avg_loss)/desired_world_size = sum(total_loss)/(desired_batch_size * desired_world_size) = sum(total_loss)/(actual_gpu_batch_size * gradient_accumulation_steps * actual_world_size) = (sum(avg_loss)/desired_world_size) * (desired_world_size / (actual_world_size * gradient_accumulation_steps))`

Or for this code: `loss = (loss * world_size_factor) / gradient_accumulation_steps`, which would be a factor 1.5 instead of 6 on my machine. Which sounds more plausible. I don't have a computer nearby now to test any of this, so I hope this text was clear enough.




Fun fact: this also tripped me up [all the way back in 2020](https://github.com/huggingface/transformers/issues/8486), you'd think remembering posting this issue would result in not making the same mistake again, haha.
